File: src/Quadrature.jl
using FastGaussQuadrature

abstract type AbstractQuadrature <: JalerkinType end;

#--- Quadrature

struct Quadrature{N} <: AbstractQuadrature
    """
    Quadrature rule with N nodes and weights of type T
    """
    nodes :: SVector{N, fp_precision()}
    weights :: SVector{N, fp_precision()}
end

@inline function get_nodes(q::Quadrature)
    q.nodes
end

@inline function get_weights(q::Quadrature)
    q.weights
end

@inline function quad(q::AbstractQuadrature, values::AbstractArray)
    """
    Compute integral of a function using a concrete quadrature rule,
    where 'values' are the values of the function on q.nodes
    """
    get_weights(q) ⋅ values
end

@inline function quad(q::AbstractQuadrature, f)
    """
    Compute integral of a function using a concrete quadrature formula
    """
    quad(q, f.(get_nodes(q)))
end

#--- Constructors for concrete quadrature rules

"""Gaussian quadrature rule on [-1,1] with n points"""
function GaussianQuadrature(n)
    nodes, weights = gausslegendre(n)
    Quadrature{n}(nodes, weights)
end

# """
# Trapezoid quadrature rule on interval [0, 1]
# """
# function TrapezoidQuadrature_01()
#     nodes = [0, 1]
#     weights = [0.5, 0.5]
#     Quadrature{2}(nodes, weights)
# end

File: py/quadrature.py
import numpy as np
from numpy.core.multiarray import array as array

class Quadrature():

    def __init__(self,nodes: np.array, weights: np.array):
        
        assert len(nodes)==len(weights), "Different dimension for nodes and weights"
        
        self.nodes=nodes
        self.weights=weights
    
    def get_nodes(self):
        return self.nodes

    def get_weights(self):
        return self.weights
    
    def quad_f(self, f):
        return np.dot(self.weights, [f(x,y) for x,y in self.nodes])
        
    
    def quad(self, val: np.array):
        return np.dot(self.weights,val)
    
#Trapezoid quadrature rule on interval [-1, 1]
class TrapezoidQuadrature1D(Quadrature):
    def __init__(self):
        self.nodes=np.array([-1,1])
        self.weights=np.array([1,1])

class TrapezoidQuadrature2D(Quadrature):
    def __init__(self):
        self.nodes=np.array([[0,0],[1,0],[0,1]])
        self.weights=np.array([1/6,1/6,1/6])

# Gauss Legendre Implement
